File: leetcode/lc-us/cpp/909.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
    int n, target;

    pair<int, int> get_position(int x) {
        int row = (x - 1) / n;
        int col;
        if (row & 1) col = (n - x % n) % n;
        else col = (x + n - 1) % n;
        return {n - row - 1, col};
    }

public:
    int snakesAndLadders(vector<vector<int>> &board) {
        n = (int) board.size(), target = n * n;
        unordered_map<int, int> dist;
        dist[1] = 0;
        queue<int> q;
        q.push(1);
        while (!q.empty()) {
            int t = q.front();
            q.pop();
            for (int move = 1; move <= 6; move++) {
                int des = t + move;
                if (des > target) break;
                auto[x, y] = get_position(des);
                if (board[x][y] != -1) {
                    des = board[x][y];
                }
                if (des == target) return dist[t] + 1;
                if (!dist.count(des)) {
                    dist[des] = dist[t] + 1;
                    q.push(des);
                }
            }
        }
        return -1;
    }
};

File: leetcode/lc-us/py/909.py
from typing import List


class Solution:
    def snakesAndLadders(self, board: List[List[int]]) -> int:
        n, m = len(board), len(board[0])

        def get_pos(idx: int) -> tuple[int, int]:
            r = n - 1 - (idx - 1) // m
            c = (idx - 1) % m
            if (r & 1) == (n & 1):
                c = m - 1 - c
            return r, c

        dist = [float('inf')] * (n * m + 1)
        dist[1] = 0
        q = [(1, 0)]
        while q:
            p, d = q[0]
            q = q[1:]
            for i in range(1, 7):
                ne = p + i
                if ne > n * m:
                    break
                x, y = get_pos(ne)
                if board[x][y] != -1:
                    ne = board[x][y]
                if dist[ne] > d + 1:
                    dist[ne] = d + 1
                    q.append((ne, d + 1))
        return dist[n * m] if dist[n * m] != float('inf') else -1
